package/MDAnalysis/analysis/Aggregation_size.py

```diff
@@ -0,0 +1,370 @@
+        super(Aggregation_size, self).__init__(trajectory)
+        # set atomgroup as a property for access in other methods
+        self.atomgroup = atomgroup
+        self._molnum = number_of_molecules
+        self._type = Type
+        self._cut_off = cut_off
+        self._monomer = no_monomer
+        self._Gyr_calc = Gyr_calc
+        self._GYR = []
+
+    def closest_atom(self, d):
+        """
+        This function calculates the distance between the closest atom
+        of two molecules.
+        """
+        self._cut_off = self._cut_off or 3.5
+        for i in range(1, self._molnum+1):
+            ii = 0
+            for j in range(i+1, self._molnum+1):
+                residue1 = u.select_atoms(f"resid {i}").positions
+                residue2 = u.select_atoms(f"resid {j}").positions
+                dists = distances.distance_array(
+                    residue1, residue2, box=u.universe.dimensions,
+                    backend='OpenMP')
+                d[i][ii] = np.amin(dists)
+                ii += 1
+
+    def center_of_mass(self, d):
+        """
+        This function calculates the distance between the COM
+        of two molecules.
+        """
+        self._cut_off = self._cut_off or 8.5
+        cm = {}
+        for x in range((self._molnum)):
+            cm[x+1] = u.select_atoms(f"resid {x+1}").center_of_mass()
+        cm_coordintes = np.array(list(cm.values()))
+        dists = distances.distance_array(
+            cm_coordintes, cm_coordintes, box=u.universe.dimensions)
+        for j in range(self._molnum):
+            ii = 0
+            for i in range(j+1, self._molnum):
+                d[j+1][ii] = dists[i][j]
+                ii += 1
+
+    def specific_atoms(self, d):
+        """
+        This function calculates the distance between the selected atom
+        of two molecules.
+        """
+        if (not(atom_1 or atom_2)):
+            raise ValueError("enter valid atom_1 and atom_2")
+        self._cut_off = self._cut_off or 8.5
+        residue1 = u.select_atoms(f"name {atom_1}").positions
+        residue2 = u.select_atoms(f"name {atom_2}").positions
+        dists = distances.distance_array(
+            residue1, residue2, box=u.universe.dimensions)
+        for j in range(self._molnum):
+            ii = 0
+            for i in range(j+1, self._molnum):
+                d[j+1][ii] = dists[i][j]
+                ii += 1
+
+    def _prepare(self):
+        """
+        Create array of zeroes as a placeholder for differnet parameters.
+        """
+        self.results = np.zeros((self.n_frames, 5))
+        self.d = [0] * self._molnum
+        i = 0
+        for it in reversed(range(1, self._molnum+1)):
+            self.d[i] = np.zeros(it * 1).reshape(it, 1)
+            i += 1
+        self.size_num = [0] * self.n_frames
+
+    def aggregation(self, k, tss, counter, m, cut_off, connected):
+        """
+        This is a subfunction that enables the code to find all
```

Comment:
Have a look at LeafletFinder in `leaflet.py` where networkx is used to do something similar. Maybe you can rewrite with networkx, which might be faster.

Reply:
self.d parameter looks like [array((n)*1), array((n-1) * 1), array((n-2) * 1), ... , array((1)*1)]. Each array consists of the distance between a molecule and the rest of the molecules. So, in the main function, provided in _single_frame, the code will go through one array (let's say the first array) and find the connected molecules to the first molecule. Then in the aggregation function, code will check if the molecules which were connected to the molecule 1 are connected to the other molecules or not. This loop will continue until there will not be any new connection. 
I read Networkx tutorial and looked to the "leaflet.py" code. Unfortunately, I could not find a similar function. Please advice me and let me know if I have to rewrite the code with Netwrokx.

Reply:
What you describe is one way of getting the connected components of a graph. In short, what the leaflet code does is that it makes a graph where each "molecule" is a node, and edges are formed when the distance between these molecules are lesser than the cut-off. Then it uses networkx to find the connected component, being the parts of the graph where each node can be accessed from any other node by following the edges.

What you do in your code is basically the same thing. Each of your "types" are different ways of building the distance matrix to build the graph.